src/qibolab/instruments/qm/config.py

```diff
@@ -48,28 +48,50 @@ def register_port(self, port):
             self.register_port(port.q)
         else:
             is_octave = isinstance(port, (OctaveOutput, OctaveInput))
+            is_fem = isinstance(port, (FEMOutput, FEMInput))
             controllers = self.octaves if is_octave else self.controllers
             if port.device not in controllers:
                 if is_octave:
                     controllers[port.device] = {}
+                elif is_fem:
+                    controllers[port.device] = {"type": "opx1000", "fems": {}}
                 else:
                     controllers[port.device] = {
                         "analog_inputs": DEFAULT_INPUTS,
                         "digital_outputs": {},
                     }
 
-            device = controllers[port.device]
+            if is_fem:
+                fems = controllers[port.device]["fems"]
+                if port.fem_number not in fems:
+                    fems[port.fem_number] = {
+                        "type": port.fem_type,
+                        "analog_inputs": DEFAULT_INPUTS,
+                        "digital_outputs": {},
+                    }
+                device = fems[port.fem_number]
+            else:
+                device = controllers[port.device]
+
             if port.key in device:
                 device[port.key].update(port.config)
             else:
                 device[port.key] = port.config
 
             if is_octave:
-                con = port.opx_port.i.device
-                number = port.opx_port.i.number
-                device["connectivity"] = con
                 self.register_port(port.opx_port)
-                self.controllers[con]["digital_outputs"][number] = {}
+                subport = port.opx_port.i
+                if isinstance(subport, (FEMOutput, FEMInput)):
+                    con = subport.device
+                    fem = subport.fem_number
+                    number = subport.number
+                    device["connectivity"] = (con, fem)
+                    self.controllers[con]["fems"][fem]["digital_outputs"][number] = {}
+                else:
+                    con = subport.device
+                    number = subport.number
```

Comment:
These are repeated in the two branches.

Reply:
Thanks, fixed in 9e510dbf17277cdea7c71a4ea1121124939b8bcc.